zrml/liquidity-mining/src/distribute_incentives_based_on_bought_shares.rs

```diff
@@ -0,0 +1,106 @@
+                    let one = T::BlockNumber::from(1u8);
+
+                    values.perpetual_incentives = values
+                        .perpetual_incentives
+                        .saturating_add(perpetual_incentives);
+                    values.total_incentives = values.total_incentives.saturating_add(incentives);
+                    values.total_shares = values.total_shares.saturating_add(bought_shares);
+                    values.participated_blocks = values.participated_blocks.saturating_add(one);
+                });
+            }
+        }
+
+        <BlockBoughtShares<T>>::remove_all();
+    }
+
+    // ZTG value of one bought share for the current block being produced. Or in other words:
+    // Determines how much a share will be worth given the amount of ZTG for liquidity
+    // mining and the total number of bought shares for the current block.
+    //
+    // `None` result means no-one purchased a share.
+    fn bought_share_value(
+        incentive: &BalanceOf<T>,
+        total_bought_shares: &BalanceOf<T>,
+    ) -> Option<BalanceOf<T>> {
+        incentive.checked_div(&total_bought_shares)
+    }
+
+    // How much incentive each market will receive
+    fn markets_incentives(
+        per_block_incentives: BalanceOf<T>,
+        curr_block: T::BlockNumber,
+    ) -> Vec<(T::MarketId, BalanceOf<T>)> {
+        let mut normalized_total = BalanceOf::<T>::from(0u8);
+        let normalized_valeus: Vec<_> = <Markets<T>>::iter()
+            .filter_map(|(market_id, period)| {
+                let normalized_value = Self::normalize_market(curr_block, period)?;
+                normalized_total = normalized_total.saturating_add(normalized_value);
+                Some((market_id, normalized_value))
+            })
+            .collect();
+        normalized_valeus
+            .into_iter()
+            .filter_map(|(market_id, normalized_value)| {
+                let incentive_ptd = calculate_perthousand(normalized_value, &normalized_total)?;
+                let incentive = calculate_perthousand_value(incentive_ptd, per_block_incentives);
+                Some((market_id, incentive))
+            })
+            .collect()
+    }
+
+    // Takes any market parameter and outputs a number that will be used as a percentage
+    // to calculate how much incentives each individual market will receive.
+    //
+    // In this case, the output is the percentage of the remaining number of blocks to stimulate
+    // early liquidity providers. For example, a market starts at 0 and ends at 10. If the current
+    // block is at 7, then the output is 30.
+    fn normalize_market(
+        curr_block: T::BlockNumber,
+        [start, end]: [T::BlockNumber; 2],
+    ) -> Option<BalanceOf<T>> {
+        let total_blocks = end.saturating_sub(start);
+        let remaining_blocks = end.saturating_sub(curr_block);
+        let ptd = calculate_perthousand(remaining_blocks, &total_blocks)?;
+        Some(perthousand_to_balance(ptd))
```

Comment:
The final result is actually `300` units (30%/300ptd of 10) instead of `3000` units (300%/3000ptd of 10).

But yes, it is per-thousand because percent incurs a greater lost of accuracy. For example, 
https://github.com/zeitgeistpm/zeitgeist/pull/122/files#diff-6d701ba4fa5e808263982c3ef607ad265031bb560ca867c28423f302da40faf4R152-R153
would be 
```rust
        // Market 0 has 42% of `per_block_incentives`
        let market_incentives_0 = 420000000;
```
with percentages.

> Is it possible that remaining_blocks can become 0 (current = end)? In that case we'd have a zero division error.

All per-thousands are calculated through the unified `calculate_perthousand` function that guards against divisions by zero -> https://github.com/zeitgeistpm/zeitgeist/pull/122/files#diff-2fa2c31f9a820cbd97164a96acb2d7648da09ec2226e576c034554d1b1b8e1cbR8-R16.

Reply:
If the remaining blocks can become zero (current = end), we should return 0% instead of a zero division error.

Reply:
Oh, I see now. Sorry for the misunderstanding and thanks for spotting this behaviour